### DISCUSSION TITLE
Add new SPAM and MALICIOUS reaction types to the protocol

### DISCUSSION THREAD

**Discussion by timjrobinson** [farcasterxyz/protocol#144] [FIP Stage 1: Ideas]
### Author

Tim - @timjrobinson

### Problem

Farcaster usage is going parabolic and along with this will come bots and malicious actors attempting to steal user funds. As this is a decentralized network we should not be relying on moderators, and instead self moderate among our social groups. 

We need some way for users to signal to other users that a post is spam (low quality, not worth reading) or malicious (harmful to interact with)

### Specification

Adds two new Reaction types, SPAM and MALICIOUS, to the protocol. Clients are free to implement these as they wish.

```
export enum ReactionType {
  ...
  /** SPAM - Flag the target cast as spam */
  SPAM = 3,
  /** MALICIOUS - Flag the target cast as malicious */
  MALICIOUS = 4,
}

export function reactionTypeFromJSON(object: any): ReactionType {
  switch (object) {
	...
    case 3:
    case "REACTION_TYPE_SPAM":
        return ReactionType.SPAM;
    case 4:
    case "REACTION_TYPE_MALICIOUS":
        return ReactionType.MALICIOUS;
	...
  }
}

export function reactionTypeToJSON(object: ReactionType): string {
  switch (object) {
    ...
    case ReactionType.SPAM:
      return "REACTION_TYPE_SPAM";
    case ReactionType.MALICIOUS:
      return "REACTION_TYPE_MALICIOUS"
    ...
  }
}
```

### Rationale

Adding these two new react types will allow users to mark messages as spam or malicious. Clients can then, using a users social graph, automatically hide content that friends have marked as spam or malicious. Hubs may also not store or rebroadcast spam or malicious content, as decided by the hub operator.  

My definitions of these two react types is:

SPAM - Casts that are not harmful, but generally a waste of time to read and adds no value to the conversation. 
MALICIOUS - Casts that contain or link to a site containing malware, wallet draining tools, malicious transactions, etc. 

The reason to have two react types is:
- Clients may want to still show spam posts, as what users deem "Quality Content" can be subjective. 
- Clients will never want to show malicious posts, as this is an objectively bad cast.

**Comment by timjrobinson** [comment]
I think it would also be useful to have:

- A react type for marking posts as explicit content (Porn, Violence, etc)
- A react type for marking illegal content (and subtypes of illegality)
- A new link type to mark a user as an impersonator. 

Though unsure if they should be added to this FIP or new ones for each, as these could be much more complex.

**Comment by varunsrin** [comment]
I'm not convinced that this will be useful to fight spam for a few reasons: 

1. Users who report a lot will get harassed 
2. Spammers or malicious actors can falsely report other people 

It would be better to start with a non-decentralized but privacy preserving reporting service and see how that works before enshrining something into the protocol

**Reply by timjrobinson** [reply]
1. This is a worry, I would like to see something privacy preserving too. 
2. This isn't an issue if this is used in a user-centric way, where clients only trust reports from people they follow. So unless you follow spammers/malicious actors they can't affect your feed.

**Comment by varunsrin** [comment]
This has been addressed by https://github.com/farcasterxyz/protocol/discussions/216
